sklego/meta.py

```diff
@@ -100,7 +152,69 @@ def __remove_groups_from_x(self, X):
         else:
             return np.delete(X, self.groups, axis=1)
 
-    def fit(self, X, y):
+    def __fit_grouped_estimator(self, X, y, value_columns, group_columns):
+        group_indices = X.groupby(group_columns).indices
+
+        try:
+            grouped_estimations = {
+                group: clone(self.estimator).fit(X.loc[indices, value_columns], y.loc[indices])
+                for group, indices in group_indices.items()
+            }
+        except KeyError:
+            raise KeyError("The indices of y and X should match")
+
+        return grouped_estimations
+
+    def __check_shrinkage_func(self):
+        """Validate the shrinkage function"""
+        group_lengths = [10, 5, 2]
+        expected_shape = np.array(group_lengths).shape
+        try:
+            result = self.shrinkage_function_(group_lengths)
+            if not isinstance(result, np.ndarray):
```

Comment:
Does not make sense to raise an Error and catch it directly. The following makes more sense (in my opinon)

```      
try:
    result = self.shrinkage_function_(group_lengths)
except Exception:         # specify the exception: ValueError, etc
     raise ValueError(f"Caught an exception while checking the shrinkage function.") from e
else:
    if not isinstance(result, np.ndarray):
        raise ValueError(f"shrinkage_function({group_lengths}) should return an np.ndarray")
    if result.shape != expected_shape:
         raise ValueError(f"shrinkage_function({group_lengths}).shape should be {expected_shape}")
```
```

Reply:
The thing with this one is that we really don't know what goes wrong in the user specified shrinkage function, so we want to catch all of these errors

Reply:
Although I do agree that `try ... except ... else` is an improvement!